### PR TITLE
Add sortColumnsBy DataFraneExt

### PR DESCRIPTION
**Pull request by zeotuan**
This was inspired by https://stackoverflow.com/questions/57821538/how-to-sort-columns-of-nested-structs-alphabetically-in-pyspark#:~:text=You%20can%20first%20flatten%20your%20DF%20with%20the%20nice%20colname.*

Here we support more flexible sorting by allowing user to provide a custom transformation and implicit ordering. 

+ add support to sorting of ArrayType(StructType) column which was not available in original implementation

### CODE REVIEW

File: src/main/scala/com/github/mrpowers/spark/daria/sql/DataFrameExt.scala [hunk 1]
```diff
@@ -407,6 +410,44 @@ object DataFrameExt {
...
+     *   // Output:
+     *   // +---+----+------+
+     *   // |age|name|salary|
+     *   // +---+----+------+
+     *   // | 30|John|2000.0|
+     *   // | 25|Jane|3000.0|
+     *   // +---+----+------+
+     * }}}
+     */
+    def sortColumnsBy[A](f: StructField => A)(implicit ord: Ordering[A]): DataFrame =
```
Comment (alfonsorr) [inline src/main/scala/com/github/mrpowers/spark/daria/sql/DataFrameExt.scala:449]: I can only see to use it with alphabetical order, if you want to allow it to use it with datatypes we should provide an instance of `Ordering[DataType]` or limit it only to order by string value.

Comment (zeotuan) [reply src/main/scala/com/github/mrpowers/spark/daria/sql/DataFrameExt.scala:449]: Sure. I think another use case would be to perform specific ordering based on some field metadata but we can allow the user the flexibility to provide their own

File: src/main/scala/com/github/mrpowers/spark/daria/sql/DataFrameExt.scala [hunk 2]
```diff
@@ -266,9 +269,9 @@ object DataFrameExt {
      * This StackOverflow answer provides a detailed description how to use flattenSchema: https://stackoverflow.com/a/50402697/1125159
      */
     def flattenSchema(delimiter: String = "."): DataFrame = {
-      val renamedCols: Array[Column] = StructTypeHelpers
+      val renamedCols = StructTypeHelpers
```
[the remark this answers is not in the document]
Comment (zeotuan) [reply src/main/scala/com/github/mrpowers/spark/daria/sql/DataFrameExt.scala:272]: Ah that was because I updated the flattenSchema return type to be Seq. Should I revert it back?

File: src/main/scala/com/github/mrpowers/spark/daria/sql/types/StructTypeHelpers.scala [hunk 3]
```diff
@@ -21,21 +21,34 @@ object StructTypeHelpers {
...
+        field.dataType match {
+          case t: StructType =>
+            acc ++ flattenSchema(t, colName)
+          case ArrayType(t: StructType, _) if flattenArrayType =>
+            acc ++ flattenSchema(t, colName)
+          case _ =>
+            acc :+ col(colName)
+        }
+    }
+  }
```
Comment (zeotuan) [inline src/main/scala/com/github/mrpowers/spark/daria/sql/types/StructTypeHelpers.scala:37]: This one is not actually completely related. But I do have an use case where I want to flatten even the Array of Struct column. Will split into another PR if you guy prefer it @alfonsorr @MrPowers

File: src/main/scala/com/github/mrpowers/spark/daria/sql/DataFrameExt.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (alfonsorr) [reply src/main/scala/com/github/mrpowers/spark/daria/sql/DataFrameExt.scala:272]: If its not related to the function create another pr so we can track when was added what. Also I know you didn't add it but I see it very complex to have a flatSchema with other flatSchema inside.
And a last thing is the method called to flat the schema, I'm not sure if you chain a array of structs with a inner column of array of structs can be flatten calling `a.b.c.d` method.

File: src/main/scala/com/github/mrpowers/spark/daria/sql/types/StructTypeHelpers.scala (hunk 3 above)
Comment (alfonsorr) [reply src/main/scala/com/github/mrpowers/spark/daria/sql/types/StructTypeHelpers.scala:37]: Didnt arrived to this comment hehe, if you can move it to other pr yup. THX!

File: src/main/scala/com/github/mrpowers/spark/daria/sql/DataFrameExt.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (zeotuan) [reply src/main/scala/com/github/mrpowers/spark/daria/sql/DataFrameExt.scala:272]: Ah yeah that's an interesting case. I'll separate this to another specific PR with some more tests

File: src/test/scala/com/github/mrpowers/spark/daria/sql/DataFrameExtTest.scala [hunk 4]
```diff
@@ -961,6 +961,389 @@ object DataFrameExtTest extends TestSuite with DataFrameComparer with SparkSessi
...
+                        )
+                      )
+                    )
+                  ),
+                  StructField(
+                    "baz",
+                    StringType,
+                    true
+                  )
+                )
```
Comment (alfonsorr) [inline src/test/scala/com/github/mrpowers/spark/daria/sql/DataFrameExtTest.scala:1027]: This is what I call a complex schema xD

**Review by alfonsorr — APPROVED** [review]

**Comment by alfonsorr** [comment]
Let me enable CI on pr first

File: src/test/scala/com/github/mrpowers/spark/daria/sql/DataFrameExtTest.scala [hunk 5]
```diff
@@ -961,6 +961,389 @@ object DataFrameExtTest extends TestSuite with DataFrameComparer with SparkSessi
...
+              "v",
+              ArrayType(ArrayType(StructType(Seq(StructField("v2", StringType, true), StructField("v1", StringType, true))))),
+              true
+            ),
+            StructField(
+              "w",
+              ArrayType(StructType(Seq(StructField("y", StringType, true), StructField("x", StringType, true)))),
+              true
+            )
+//            StructField(
```
Comment (zeotuan) [inline src/test/scala/com/github/mrpowers/spark/daria/sql/DataFrameExtTest.scala:1224]: @alfonsorr The problem is it does not work yet with this last case which I commented out.

I am thinking making ArrayType Sorting an optional setting and add in documentation that 
we not yet support sorting of Array(Struct(Array)) 

Currenty it work for:
- NDimensional Array
- Struct(Array(Struct))

Not supported:
- Array(Struct(Array))

Comment (zeotuan) [reply src/test/scala/com/github/mrpowers/spark/daria/sql/DataFrameExtTest.scala]: I Managed to get it to work with this case so ignore this

File: src/main/scala/com/github/mrpowers/spark/daria/sql/types/StructTypeHelpers.scala [hunk 6]
```diff
@@ -38,6 +39,55 @@ object StructTypeHelpers {
     })
   }
 
+  private def schemaToSortedSelectExpr[A](schema: StructType, f: StructField => A, baseField: String = "")(implicit ord: Ordering[A]): Seq[Column] = {
+    def handleNestedType(t: DataType, name: String, outerCol: Column): Column =
```
Comment (zeotuan) [inline src/main/scala/com/github/mrpowers/spark/daria/sql/types/StructTypeHelpers.scala:43]: @alfonsorr finally managed to get it to working with Array(Struct(Array(Struct))) cases By rewrite completely using transform instead of name concat. But it resulted in this mess will try a bit more if I can deobfuscate it
